### ISSUE TITLE
ValueTask support

### ISSUE THREAD

**Issue by splitice** [StackExchange/StackExchange.Redis#2344]
When investigating performant ways to iterate a large database we discovered that there is definately room for decreasing Task overhead.

How? Well benchmarks showed 8% of our CPU being used in setting the task result (along with other task related overheads).

We verified this by testing a blocking version of the code against a non-blocking version and was able to see a substantial improvement

```
        private async IAsyncEnumerable<HashEntry[]> GetConntracksRawAsync(Func<Task<IEnumerable<IServer>>> redisServerSelect)
        {
            var databaseId = _redis.Database;
            var redisServers = await redisServerSelect();
            foreach (var server in redisServers)
            {
                await foreach (var key in server.KeysAsync(databaseId))
                {
                    var dict = await _redis.HashGetAllAsync(key);
                    yield return dict.Length == 0 ? null : dict;
                }
            }
        }
```
11s to complete

```
        private async IAsyncEnumerable<HashEntry[]> GetConntracksRaw(Func<Task<IEnumerable<IServer>>> redisServerSelect)
        {
            var databaseId = _redis.Database;
            var redisServers = await redisServerSelect();
            foreach (var server in redisServers)
            {
                foreach (var key in server.Keys(databaseId))
                {
                    var dict = _redis.HashGetAll(key);
                    yield return dict.Length == 0 ? null : dict;
                }
            }
        }
```
8s to complete over 100k items

ValueTask would be very close to 8s. We have been using ValueTask extensively ourselves and have seen Task overheads drop substantially (to 3-10% of what was previously used by `Task`).

Related:
- #1855
- #1295

**Comment by NickCraver** [comment]
What is the ask here? The calls to Redis are _actually_ async, so a ValueTask does not have the same benefits as when "it's probably coming back without going async", because when it's async ValueTask _also_ creates a Task, meaning there's _more_ overhead, not less. Also: it'd be a breaking change to move all APIs (as noted in other issues).

Also note that measuring completion time for this benchmark isn't a good benchmark to go by for performance. You're testing a certain latency and utilizing the thread pool (which would cause starvation under load) for the synchronous variant. Can it sometimes be faster in isolation? Yes. Are there other costs that benchmark is not demonstrating? Also yes :)

So overall the ask I have is: what is the request here?

**Comment by splitice** [comment]
> What is the ask here? The calls to Redis are _actually_ async, so a ValueTask does not have the same benefits as when "it's probably coming back without going async"

I'm pretty sure there is enough people using redis in near instantantaneous applications (I'd argue most) as opposed to massively high latency applications.

ValueTask works best when the next item is pre-fetched or when fetching by pages. I'm not sure if its possible to get fetching by pages with the current API regardless but pre-fetching the hash of the next item is definately possible and would give you Taskless iteration.

ValueTask would have the advantage of being safe to use in TPL but also (nearly) as performant as the Sync api... Thats the entire point for ValueTask. PolledValueTask is typically even closer (although thats not an initial target goal most likely).

> Also: it'd be a breaking change to move all APIs (as noted in other issues).

I'm not suggesting replacement. I'm suggesting the same thing every other issue regarding this has, an additional set of APIs.

Also you could probably make use of ValueTask internally (I'm sure there are applications for it).

> You're testing a certain latency and utilizing the thread pool (which would cause starvation under load) for the synchronous variant

Yep, but not with a ValueTask API if one existed. Can't benchmark with something that doesnt exist, however ValueTask when awaited on an already existing 

> Are there other costs that benchmark is not demonstrating?

Program: A function call from IEnumerable<>.ToList()/.ToListAsync() and StopWatch.... aka nothing unreasonable
System: TCP encapsulation of redis data over localhost, redis is not pegged (~10% of one core)

```
        [Theory]
        [InlineData(100000)]
        public async Task BenchGetConntracksRaw(int nConntracks)
        {
            var db = _redis.GetDatabase(6);
            var server = _redis.GetServer(await db.IdentifyEndpointAsync());

            await server.FlushDatabaseAsync(db.Database);

            // [.. fill db ]

            Stopwatch sw = new Stopwatch();
            sw.Start();
            var cts = ctDb.GetConntracksRaw().ToList();
            sw.Stop();
            Log.Logger.Information("Elapsed {time}", sw.Elapsed);
        }
```

**Comment by NickCraver** [comment]
> I'm pretty sure there is enough people using redis in near instantantaneous applications (I'd argue most) as opposed to massively high latency applications.

How are you defining this? Massive high latency here compared to memory is "hitting the network at all", even if it's 0.1ms. That's most users.

> ValueTask works best when the next item is pre-fetched or when fetching by pages. I'm not sure if its possible to get fetching by pages with the current API regardless but pre-fetching the hash of the next item is definately possible and would give you Taskless iteration.
ValueTask would have the advantage of being safe to use in TPL but also (nearly) as performant as the Sync api... Thats the entire point for ValueTask. PolledValueTask is typically even closer (although thats not an initial target goal most likely).

This would apply to very few APIs, but possible. It's not the same safety as Task (e.g. multi-await), which is something users would hit. Saying it's nearly as performant as the sync API is highly subjective though, as a) it's very constrained to paged things (only a hand full of calls at best) and b) thread pool semantics which I cannot overstate here. The vast majority of our timeout issues from users are around apps doing this and us being the ones alerting people to it. Overall: you're testing with nearly 0 latency and that's not most users, the factors change greatly when the majority of time in any loop iteration is network latency.

> Also you could probably make use of ValueTask internally (I'm sure there are applications for it).

We do this where it makes sense today, but feel free to suggest more internal usages where it improves something!

> Yep, but not with a ValueTask API if one existed. Can't benchmark with something that doesnt exist, however ValueTask when awaited on an already existing

Sure, but this could be prototyped and benchmarked. That'd be required here to justify adding confusion to the API.

**Comment by mgravell** [comment]
Nick's already covered the main point here, which is that the CPU overhead isn't in "task Vs ValueTask" - it is in the state machine and thread-pool switching, and in the asynchronous case (which would always apply here): that *doesn't change between task and valuetask*.

The specific advantage of ValueTask isn't really about CPU - it is about allocations. There are 3 different modes for ValueTask:

- wrapping a synchronous value (doesn't apply here)
- wrapping a Task
- via IValueTaskSource, amortizing allocs

Now, that 3rd option does apply here, and in theory we could indeed expose values as ValueTask backed via IValueTaskSource - unfortunately, that 3rd option didn't exist when we created the API shape. Switching the API to ValueTask would be a fundamentally breaking change, *plus* there's the multi-await problem.

That said: I am open to experimenting. If there's a specific workload you have a concrete measure for (perhaps the above, although a well-known data set would be key, too) - we can *check* whether there's any advantage to be had. But in reality: I would *not* expect any major CPU advantage if we did this. What I would expect is a measurable decrease in allocations, with very little CPU impact.
